File: native/YMC/Classes/Manifest.swift
//
//  Manifest.swift
//  YMC
//
//  Created by Максим Карелов on 20/02/2019.
//  Copyright © 2019 Maksim Karelov. All rights reserved.
//

import Foundation
import Files

enum WriteManifestStatus {
    case updated
    case skipped
    case error
}

struct Manifest: Codable, Equatable {
    let name: String
    let description: String
    let path: String
    let type: String
    var allowed_extensions: [String]? = nil
    var allowed_origins: [String]? = nil

    init(name: String, description: String, path: String, type: String) {
        self.name = name
        self.description = description
        self.path = path
        self.type = type
    }

    static func == (a: Manifest, b: Manifest) -> Bool {
        return (
            a.name == b.name &&
            a.description == b.description &&
            a.path == b.path &&
            a.type == b.type &&
            (
                (a.allowed_extensions != nil && b.allowed_extensions != nil && a.allowed_extensions!.elementsEqual(b.allowed_extensions!)) ||
                (a.allowed_origins != nil && b.allowed_origins != nil && a.allowed_origins!.elementsEqual(b.allowed_origins!))
            )
        )
    }
}

private let paths = [
    "\(Folder.home.path)Library/Application Support/Mozilla",
    "\(Folder.home.path)Library/Application Support/Google/Chrome",
    "\(Folder.home.path)Library/Application Support/Chromium",
    "\(Folder.home.path)Library/Application Support/com.operasoftware.Opera",
    "\(Folder.home.path)Library/Application Support/Yandex/YandexBrowser",
    "\(Folder.home.path)Library/Application Support/Vivaldi"
]

private func createManifest(appPath: String, isChromeBasedBrowser: Bool) -> Manifest {
    var manifest = Manifest(
        name: "ymc",
        description: "Control Yandex.Music from any window in MacOS",
        path: appPath,
        type: "stdio"
    )

    if isChromeBasedBrowser {
        manifest.allowed_origins = ["chrome-extension://mhjachgjcpppedbkfaajhodfgiilbmci/"]
    } else {
        manifest.allowed_extensions = ["ymc@karelov.info"]
    }

    return manifest
}

private func writeManifest(_ path: String) -> WriteManifestStatus {
    guard let appPath = Bundle.main.executablePath else { return .error }

    let manifest = createManifest(appPath: appPath, isChromeBasedBrowser: !path.contains("Mozilla"))
    let encoder = JSONEncoder()

    encoder.outputFormatting = .prettyPrinted

    guard let manifestData = try? encoder.encode(manifest) else { return .error }
    guard let manifestDataString = String(data: manifestData, encoding: .utf8)?
        .replacingOccurrences(of: "\\/", with: "/")
        .replacingOccurrences(of: " : ", with: ": ") else { return .error }
    guard let manifestDataFixed = manifestDataString.data(using: .utf8) else { return .error }
    guard let browserFolder = try? Folder(path: path) else { return .skipped }

    var folder: Folder

    if let targetFolder = try? browserFolder.subfolder(named: "NativeMessagingHosts") {
        folder = targetFolder
    } else if let targetFolder = try? browserFolder.createSubfolder(named: "NativeMessagingHosts") {
        folder = targetFolder
    } else {
        return .skipped
    }

    var targetFile: File?

    if let file = try? folder.file(named: "ymc.json") {
        guard let existingManifestData = try? file.read() else { return .error }

        if let existingManifest = try? JSONDecoder().decode(Manifest.self, from: existingManifestData), existingManifest == manifest {
            return .skipped
        }

        targetFile = file
    } else {
        targetFile = try? folder.createFile(named: "ymc.json")
    }

    if targetFile == nil {
        return .error
    }

    do {
        try targetFile!.write(data: manifestDataFixed)
    } catch {
        return .error
    }

    return .updated
}

func processManifests() -> WriteManifestStatus {
    let results = paths.map { writeManifest($0) }

    if results.contains(.error) {
        return .error
    } else if results.contains(.updated) {
        return .updated
    }

    return .skipped
}

File: src/main/kotlin/info/karelov/ymc/classes/Manifest.kt
package info.karelov.ymc.classes

import com.beust.klaxon.JsonObject
import com.beust.klaxon.Klaxon
import com.beust.klaxon.Parser
import info.karelov.ymc.Main
import java.io.File
import java.lang.StringBuilder
import kotlin.Exception

enum class WriteManifestStatus {
    Updated,
    Skipped,
    Error
}

data class Manifest(
    val name: String,
    val description: String,
    val path: String,
    val type: String
) {
    @Suppress("PropertyName")
    var allowed_extensions: Array<String>? = null
    @Suppress("PropertyName")
    var allowed_origins: Array<String>? = null

    override fun equals(other: Any?): Boolean {
        if (other !is Manifest) {
            return false
        }

        return (
            this.name == other.name &&
            this.description == other.description &&
            this.path == other.path &&
            this.type == other.type &&
            (
                (this.allowed_extensions != null && other.allowed_extensions != null && this.allowed_extensions!!.contentEquals(
                    other.allowed_extensions!!
                )) ||
                (this.allowed_origins != null && other.allowed_origins != null && this.allowed_origins!!.contentEquals(
                    other.allowed_origins!!
                ))
            )
        )
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + description.hashCode()
        result = 31 * result + path.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + (allowed_extensions?.contentHashCode() ?: 0)
        result = 31 * result + (allowed_origins?.contentHashCode() ?: 0)
        return result
    }
}

private val USER_HOME = System.getProperty("user.home")
private val paths = arrayOf(
    "$USER_HOME/.config/google-chrome",
    "$USER_HOME/.config/google-chrome-beta",
    "$USER_HOME/.config/chromium",
    "$USER_HOME/.config/opera",
    "$USER_HOME/.config/yandex-browser",
    "$USER_HOME/.config/yandex-browser-beta",
    "$USER_HOME/.mozilla"
)

private fun createManifest(appPath: String, isChromeBasedBrowser: Boolean): Manifest {
    val manifest = Manifest(
        name = "ymc",
        description = "Control Yandex.Music from any window",
        path = appPath,
        type = "stdio"
    )

    if (isChromeBasedBrowser) {
        manifest.allowed_origins = arrayOf("chrome-extension://mhjachgjcpppedbkfaajhodfgiilbmci/")
    } else {
        manifest.allowed_extensions = arrayOf("ymc@karelov.info")
    }

    return manifest
}

private fun writeManifest(path: String): WriteManifestStatus {
    val isFirefox = path.contains(".mozilla")
    val nativeMessagingDirName = if (isFirefox) {
        "native-messaging-hosts"
    } else {
        "NativeMessagingHosts"
    }
    val appPath = File(Main::class.java.protectionDomain.codeSource.location.toURI()).absolutePath ?: return WriteManifestStatus.Error
    val manifest = createManifest(appPath, !isFirefox)
    val manifestDataUgly = Klaxon().toJsonString(manifest)
    val manifestData = (Parser.default().parse(StringBuilder(manifestDataUgly)) as JsonObject).toJsonString(true)

    if (manifestData == "null") {
        return WriteManifestStatus.Error
    }

    val browserDir = File(path)
    val folder: String

    if (!browserDir.exists()) {
        return WriteManifestStatus.Skipped
    } else {
        val nativeMessagingDir = File("$path/$nativeMessagingDirName")

        folder = if (nativeMessagingDir.exists()) {
            nativeMessagingDir.absolutePath
        } else {
            try {
                nativeMessagingDir.mkdir()
            } catch (e: SecurityException) {
                return WriteManifestStatus.Skipped
            }

            nativeMessagingDir.absolutePath
        }
    }

    val targetFile = File("$folder/ymc.json")
    val writeToFile: Boolean

    writeToFile = if (targetFile.exists()) {
        try {
            val existsManifestData = targetFile.readText()

            if (existsManifestData == manifestData) {
                return WriteManifestStatus.Skipped
            }

            true
        } catch (e: Exception) {
            true
        }
    } else {
        true
    }

    if (writeToFile) {
        try {
            targetFile.writeText(manifestData)
        } catch (e: Exception) {
            return WriteManifestStatus.Error
        }
    } else {
        return WriteManifestStatus.Skipped
    }

    return WriteManifestStatus.Updated
}

fun processManifests(): WriteManifestStatus {
    val results = paths.map { writeManifest(it) }

    if (results.contains(WriteManifestStatus.Error)) {
        return WriteManifestStatus.Error
    } else if (results.contains(WriteManifestStatus.Updated)) {
        return WriteManifestStatus.Updated
    }

    return WriteManifestStatus.Skipped
}
